docs/architecture/adr-042-state-sync.md

```diff
@@ -0,0 +1,209 @@
+alleged chunk number. This could make it difficult for nodes to figure
+out which parts of the tree they have and which parts they still need.
+Hence, to go with this chunking approach to state sync, information
+about how state is chunked would need to be hashed into the state
+itself, or into the block header, so that chunk numbers could be
+directly verified.
+
+An alternative design was suggested by Jae Kwon in
+[#3639](https://github.com/tendermint/tendermint/issues/3639) where chunking
+happens lazily and in a dynamic way: nodes request key ranges from their peers,
+and peers respond with some subset of the
+requested range and with notes on how to request the rest in parallel from other
+peers. Unlike chunk numbers, keys can be verified directly. And if some keys in the
+range are ommitted, proofs for the range will fail to verify.
+This way a node can start by requesting the entire tree from one peer,
+and that peer can respond with say the first few keys, and the ranges to request
+from other peers.
+
+#### Eager StateSync
+Warp Sync as implemented in Parity
+["Warp Sync"](https://wiki.parity.io/Warp-Sync-Snapshot-Format.html) to rapidly
+download both blocks and state snapshots from peers. Data is carved into ~4MB
+chunks and snappy compressed. Hashes of snappy compressed chunks are stored in a
+manifest file which co-ordinates the state-sync. Obtaining a correct manifest
+file seems to require an honest majority of peers. This means you may not find
+out the state is incorrect until you download the whole thing and compare it
+with a verified block header.
+
+A similar solution was implemented by Binance in
+[#3594](https://github.com/tendermint/tendermint/pull/3594)
+based on their initial implementation in
+[PR #3243](https://github.com/tendermint/tendermint/pull/3243)
+and [some learnings](https://docs.google.com/document/d/1npGTAa1qxe8EQZ1wG0a0Sip9t5oX2vYZNUDwr_LVRR4/edit).
+Note this still requires the honest majority peer assumption.
+One major advantage of the warp-sync approach is that chunks can be compressed before sending.
+
+Ideally State Sync in Tendermint would not require an honest majority of peers, and instead
+would rely on the standard light client security, meaning state chunks would be
+verifiable against the merkle root of the state. Such a design for tendermint
+was originally tracked in [#828](https://github.com/tendermint/tendermint/issues/828).
+
+### Analysis of Lazy vs Eager
+
+Lazy vs Eager have more in common than they differ. They all require
+reactors on the tendermint side, a set of ABCI messages and a method for
+serializing/deserializing snapshots facilitated by a SnapshotFormat.
+
+The biggest difference between Lazy and Eager proposals is in the
+read/write patterns necessitated by when serving a snapshot chunk.
+Specifically, Lazy State Sync random reads to the underlying data
+structure while Eager can optimize for sequential reads.
```

Comment:
Heh, I think this might only be true for poorly implemented Merkle trees (ie. most of them :P). I believe this might be what Matt was alluding to here: https://github.com/tendermint/tendermint/issues/3639#issuecomment-491531005

Reply:
Yeah in the case of [alternative trees](https://github.com/nomic-io/merk) the implementation might provide good range scan characteristics which would alleviate any need to materialize a snapshot on disk. Since we've taken the approach to define a set of expectations by way of interface specification and push as many material concerns down to the application, we need not specify where/how the snapshots are stored. I think/hope that our current design supports the ability for an ABCI application to serve chunk requests directly from the tree if they so chose, we only insist that the the indexing of chunks be consistent across peers.